File: Pilasarreglos/pilasarre.cpp
#include <iostream>
#include <vector>

class PilaArreglo {
private:
    std::vector<int> stack;
    int maxSize;
    int top;

public:
    PilaArreglo(int tamanoMax) {
        this->maxSize = tamanoMax;
        this->stack.resize(tamanoMax);
        this->top = -1;
    }

    bool isEmpty() {
        return this->top == -1;
    }

    bool isFull() {
        return this->top == this->maxSize - 1;
    }

    void push(int data) {
        if (this->isFull()) {
            std::cout << "Error: Stack Overflow" << std::endl;
            return;
        }
        this->top++;
        this->stack[this->top] = data;
    }

    int pop() {
        if (this->isEmpty()) {
            std::cout << "Error: Stack Underflow" << std::endl;
            return -1;
        }
        int data = this->stack[this->top];
        this->top--;
        return data;
    }

    int peek() {
        if (this->isEmpty()) {
            std::cout << "Pila vacía" << std::endl;
            return -1;
        }
        return this->stack[this->top];
    }
};

int main() {
    PilaArreglo pila(100);
    pila.push(10);
    pila.push(20);
    pila.push(30);

    std::cout << "Elemento superior: " << pila.peek() << std::endl;
    std::cout << "Extrae elemento: " << pila.pop() << std::endl;
    std::cout << "Nuevo elemento superior: " << pila.peek() << std::endl;

    return 0;
}

File: Pilasarreglos/Pilasarre.cs
using System;

public class PilaArreglo
{
    private int[] stack;
    private int maxSize;
    private int top;

    public PilaArreglo(int tamanoMax)
    {
        this.maxSize = tamanoMax;
        this.stack = new int[this.maxSize];
        this.top = -1;
    }

    public bool IsEmpty()
    {
        return this.top == -1;
    }

    public bool IsFull()
    {
        return this.top == this.maxSize - 1;
    }

    public void Push(int data)
    {
        if (this.IsFull())
        {
            Console.WriteLine("Error: Stack Overflow");
            return;
        }
        this.top++;
        this.stack[this.top] = data;
    }

    public int Pop()
    {
        if (this.IsEmpty())
        {
            Console.WriteLine("Error: Stack Underflow");
            return -1;
        }
        int data = this.stack[this.top];
        this.top--;
        return data;
    }

    public int Peek()
    {
        if (this.IsEmpty())
        {
            Console.WriteLine("Pila vacía");
            return -1;
        }
        return this.stack[this.top];
    }
}

public class pilaasrre
{
    public static void Main(string[] args)
    {
        PilaArreglo pila = new PilaArreglo(100);
        pila.Push(10);
        pila.Push(20);
        pila.Push(30);

        Console.WriteLine($"Elemento superior: {pila.Peek()}");
        Console.WriteLine($"Extrae elemento: {pila.Pop()}");
        Console.WriteLine($"Nuevo elemento superior: {pila.Peek()}");
    }
}